docs/src/proposals/traced-transaction-construction.md

```diff
@@ -0,0 +1,247 @@
+    amount: u64,             
+    expected_decimals: Option<u8>,    
+) -> ProgramResult {
+    let counter_address = Pubkey::find_program_address(
+        &[  
+            &program_id.to_bytes(),
+            "counter",
+        ],
+        program_id,
+    );
+   let transfer_counter = access_as_unsigned(counter_address, READ_WRITE)?;
+   let clock_sysvar = access_as_unsigned(sysvars::clock::id(), READ_WRITE)?;
+   let today = yyyymmdd(clock_sysvar);
+   if today > transfer_counter.yyyymmdd {
+      transfer_counter.yymmdd = today;
+      transfer_counter.count = 0;
+   } else if transfer_counter > 10 {
+      return Err("daily quote of transfer is reached");
+   }
+
+   transfer_counter.count += 1;
+   ...
+}
+```
+
+### example: sharding
+
+
+```rust
+pub fn process_place_order(
+  program_id: &Pubkey,     
+  market_id: &Pubkey,
+  order: &Pubkey,
+) {
+  let sharded_event_queue_address = let counter_address = Pubkey::find_program_address(
+      &[  
+          &market_id.to_bytes(),
+          &[ourder_to_bytes()[0] & 0xf0],
+      ],
+      program_id,
+  );
+  let event_queue = access_as_unsigned(sharded_event_queue_address, READ_WRITE);  
+
+  // crank tx will merge all event queues later
+}
+```
+
+### example: instruction versioning across redeploy
+
+```rust
+pub fn check_instruction_version_and_load_data() {
+   // this is newly added after the redeploy
+   if access_as_unsigned(declare!("OldVersion111111111111111111111111111111111111"), READ).is_ok() {
+     // it seems that this transaction is construction by simulation with pre-redeploy program executable.
+     return Err("we changed internal data loading implementation. to avoid unexpected malfunction retry");
+   }
+
+   // this is changed from OldVersion1111
+   access_as_unsigned(declare!("NewVersion111111111111111111111111111111111111"), READ)?
+
+   access_as_unsigned(...);
+}
+```
+
+
+## RPC changes
+
+`pre_balances` and `post_balances` needs to be updated?
+
+### sample return from `simulateTransaction` (or define `traceTransaction` for the compatibility??):
+
+```yml
+# these somewhat verbose formatting is for future expansion
+transaction:
+  instructions
+    accesses:
+      Bm9UhRBysjeT35ukzriheHKr76yiM1fPyEBMMw2yT9v1: {read: true, write: false}      
+      J8H1Gsvnm2CY5oVa1zw1LHHbfvtone8ZGsVBzjkkPbpy: {read: true, write: true}      
+    instructions:
+      accesses: []
+      instruction_type: spl_token_transfer
+      instructions: [...]
+  accesses: [...]
+  pre_balances/pre_token_balances
+  post_balances/post_token_balances
+  token_token_approvals: ???
+```
+
+
+
+also, page-indexed account can be incorporated to find compact combination of existing account indexes to match the required account list
+
+## downside
+
+race condition around program deploy (= possible account loading change) => just retry also seek to the ix versioning for perfection
+
+(traced program just load dummy addresses for tx versionsing for completeness)
```

Comment:
also, I have to embed this pr comment into doc: https://github.com/solana-labs/solana/pull/17796#issuecomment-885060430